### PR TITLE
2882 -added A26 logic

### PR DESCRIPTION
**Pull request by slamothe99**
## Pre-review checklist

I have ensured that:

- [ ] Code runs successfully in the development environment.
- [ ] New code is tested to prove that it meets the specificaiton. If not, justify reasoning.
- [ ] New code is documented using [type hinting](https://realpython.com/lessons/type-hinting/) and following the [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) docstring format.
- [ ] New code follows the project naming standards.

### CODE REVIEW

File: cishouseholds/pipeline/high_level_transformations.py [hunk 1]
```diff
@@ -2529,7 +2533,197 @@ def union_dependent_cleaning(df):
...
+    # 3
+    for col in ["think_had_covid_contacted_nhs", "think_had_covid_admitted_to_hospital", "other_covid_infection_test"]:
+        df = df.withColumn(
+            col,
+            F.when(
+                (~F.col("think_had_covid").eqNullSafe("Yes"))
+                & (F.col("other_covid_infection_test_results").isNull())
+                & (F.col("survey_response_dataset_major_version") == 0)
+                & (
+                    reduce(add, [F.when(F.col(c).isin(["Yes", "Positive"]), 1).otherwise(0) for c in covid_test_cols])
```
Comment (murphc1) [inline cishouseholds/pipeline/high_level_transformations.py:2576]: The result for other_covid_infection_test are mapped to: 
[0] Any tests negative, but none positive; **[1] One or more positive test(s);** [2] Waiting for all results; [9] All Tests failed
So depending on whether this comes after the mapping this might need to be changed to the bold string or '1'.

File: cishouseholds/pipeline/high_level_transformations.py [hunk 2]
```diff
@@ -2529,7 +2533,197 @@ def union_dependent_cleaning(df):
...
+                None,
+            ).otherwise(F.col(col)),
+        )
+    # 4
+    df = assign_column_value_from_multiple_column_map(
+        df,
+        "think_had_covid",
+        [
+            "No",
+            [0, "No", None, ["No", None], 0, None, "Yes", "Negative"],
```
Comment (murphc1) [inline cishouseholds/pipeline/high_level_transformations.py:2588]: See previous point other_covid_infection_test need to be changed to the string or '0' depending on whether this happens before or after mapping: 
[0] Any tests negative, but none positive;

File: cishouseholds/pipeline/high_level_transformations.py [hunk 3]
```diff
@@ -2529,7 +2533,197 @@ def union_dependent_cleaning(df):
...
+        ],
+        [*think_had_covid_cols, "other_covid_infection_test", "other_covid_infection_test_results"],
+    )
+    # 5
+    df = assign_column_value_from_multiple_column_map(
+        df,
+        "think_had_covid",
+        ["Yes", ["No", "Yes", "Positive"]],
```
Comment (murphc1) [inline cishouseholds/pipeline/high_level_transformations.py:2596]: Same as before. 
The result for other_covid_infection_test are mapped to:
[0] Any tests negative, but none positive; **[1] One or more positive test(s);** [2] Waiting for all results; [9] All Tests failed
So depending on whether this comes after the mapping this might need to be changed to the bold string or '1'.

File: cishouseholds/pipeline/high_level_transformations.py [hunk 4]
```diff
@@ -2529,7 +2533,197 @@ def union_dependent_cleaning(df):
...
+        ["think_had_covid", "other_covid_infection_test", "other_covid_infection_test_results"],
+    )
+    # 6
+    for col in covid_test_cols:
+        df = df.withColumn(
+            col,
+            F.when(
+                (all_columns_values_in_list(covid_test_cols, ["No", "Negative", None]))
```
Comment (murphc1) [inline cishouseholds/pipeline/high_level_transformations.py:2604]: Same as previous point - other_covid_infection_test need to be changed to the string or '0' depending on whether this happens before or after mapping:
[0] Any tests negative, but none positive;

File: cishouseholds/pipeline/high_level_transformations.py [hunk 5]
```diff
@@ -2529,7 +2533,197 @@ def union_dependent_cleaning(df):
...
+        [*think_had_covid_cols, "other_covid_infection_test", "other_covid_infection_test_results"],
+    )
+    # 5
+    df = assign_column_value_from_multiple_column_map(
+        df,
+        "think_had_covid",
+        ["Yes", ["No", "Yes", "Positive"]],
+        ["think_had_covid", "other_covid_infection_test", "other_covid_infection_test_results"],
+    )
+    # 6
```
Comment (murphc1) [inline cishouseholds/pipeline/high_level_transformations.py:2599]: This is for dataset == 0 only.

File: cishouseholds/pipeline/high_level_transformations.py [hunk 6]
```diff
@@ -2529,7 +2533,197 @@ def union_dependent_cleaning(df):
...
+        df,
+        "think_had_covid",
+        [
+            "No",
+            [0, "No", None, ["No", None], 0, None, "Yes", "Negative"],
+        ],
+        [*think_had_covid_cols, "other_covid_infection_test", "other_covid_infection_test_results"],
+    )
+    # 5
+    df = assign_column_value_from_multiple_column_map(
```
Comment (murphc1) [inline cishouseholds/pipeline/high_level_transformations.py:2593]: This is for dataset ==0 only

File: cishouseholds/pipeline/high_level_transformations.py [hunk 7]
```diff
@@ -2529,7 +2533,197 @@ def union_dependent_cleaning(df):
...
+            F.when(
+                (F.col(col) == "No")
+                & (F.col("think_had_covid") == "No")
+                & (F.col("think_had_covid_onset_date").isNull())
+                & (F.col("sympt_covid_count") == 0)
+                & (F.col("survey_response_dataset_major_version") == 1)
+                & (
+                    reduce(
+                        add,
+                        [F.when(F.col(c).isin(["No", "Negative", None]), 1).otherwise(0) for c in hospital_covid_cols],
```
Comment (murphc1) [inline cishouseholds/pipeline/high_level_transformations.py:2638]: As the noted variables can also be 'Any tests negative, but none positive' this needs to be added to this clause

File: cishouseholds/pipeline/high_level_transformations.py [hunk 8]
```diff
@@ -2529,7 +2533,197 @@ def union_dependent_cleaning(df):
...
+        df = df.withColumn(
+            col,
+            F.when(
+                (F.col(col).eqNullSafe("No"))
+                & (F.col("think_had_covid").isNull())
+                & (count_no <= 2)
+                & (F.col("other_covid_infection_test_results").isNull())
+                & (F.col("survey_response_dataset_major_version") == 0)
+            ),
+        )
```
Comment (murphc1) [inline cishouseholds/pipeline/high_level_transformations.py:2666]: This is #10 (correcting for my sanity when checking against STATA logic on confluence)

File: cishouseholds/pipeline/high_level_transformations.py [hunk 9]
```diff
@@ -2529,7 +2533,197 @@ def union_dependent_cleaning(df):
...
+    # 10
+    df = assign_column_value_from_multiple_column_map(
+        df,
+        "other_covid_infection_test_results",
+        [None, ["Negative", None, None, None, None, "No", 0]],
```
Comment (murphc1) [inline cishouseholds/pipeline/high_level_transformations.py:2671]: Change to 'Any tests negative, but none positive'

File: cishouseholds/pipeline/high_level_transformations.py [hunk 10]
```diff
@@ -2529,7 +2533,197 @@ def union_dependent_cleaning(df):
...
+            "other_covid_infection_test_results",
+            "think_had_covid",
+            "think_had_covid_onset_date",
+            "think_had_covid_contacted_nhs",
+            "think_had_covid_admitted_to_hospital",
+            "other_covid_infection_test",
+            "survey_response_dataset_major_version",
+        ],
+    )
+    # 11
```
Comment (murphc1) [inline cishouseholds/pipeline/high_level_transformations.py:2682]: Might be me not understanding this function but how does this replace think_had_covid to NULL if these conditions are met?

File: cishouseholds/pipeline/high_level_transformations.py [hunk 11]
```diff
@@ -2529,7 +2533,197 @@ def union_dependent_cleaning(df):
...
+            & (F.col("think_had_covid_onset_date_string").isNull())
+            & (F.col("survey_response_dataset_major_version") == 0)
+        ),
+    )
+    # 12
+    for col in covid_test_cols:
+        df = assign_column_value_from_multiple_column_map(
+            df,
+            col,
+            [None, [["Negative", None], "No", None, [None, "No"], [None, "No"], [None, "No"], 0, 0]],
```
Comment (murphc1) [inline cishouseholds/pipeline/high_level_transformations.py:2699]: Change to 'Any tests negative, but none positive'

**Review by murphc1 — CHANGES_REQUESTED** [review]
There are some minor changes needed (particularly with the mapping of other_covid_infection_test_result) but in general looks great 👍

**Review by murphc1 — APPROVED** [review]
